This commit adds support for a configuration parameter

josephhardinee opened the pull request:
This commit adds support for a configuration parameter to each of the cluster types in jobqueue. It does this by adding a configuration parameter that is used to override the scheduler_name class variable.
 I also demoted scheduler_name to an instance variable as you could now theoretically have multiple types of PBS clusters.

The form of usage of this is as suggested by @mrocklin. 

```python
cluster = PBSCluster(configuration='my_pbs') 
```
Will draw from the my_pbs dictionary in the yaml file rather than the normal pbs one. When called without a parameter, it still defaults to pbs. I made a similar change for all of the cluster type initiators.

I tested it on the pytests, as well as the docker tests for PBS. All passed except one at the end that I think is just a docker error with weird networking setups. 

Let me know if this looks acceptable or if you'd like any other changes.

mrocklin commented:
Thanks for doing this @josephhardinee !

Some questions/comments:

1.  Should we drop the `scheduler_name` attribute entirely?  Do we use it for anything else?  (This is more a question for other maintainers)
2.  Can we include a small test for this?  It might look like the following:

    ```python
    def test_configuration_name():
        conf = {'queue': 'foo'}        
        with dask.config.set({'jobqueue.my_custom_configuration': conf}):
            cluster = SGECluster(configuration='my_custom_configuration')
            assert cluster.queue == 'foo'
    ```
3.  Is there a better name than `configuration`?  This might be confused with all of the key-value pairs in the configuration instead.

guillaumeeb commented on dask_jobqueue/lsf.py:
```diff
@@ -50,9 +50,10 @@ class LSFCluster(JobQueueCluster):
-    def __init__(self, queue=None, project=None, ncpus=None, mem=None, walltime=None, job_extra=None, **kwargs):
+    def __init__(self, queue=None, project=None, ncpus=None, mem=None, walltime=None, job_extra=None,
+                 configuration='lsf', **kwargs):
+        self.scheduler_name = configuration
```
I believe that for clarity, `self.scheduler_name` attribute should be renamed to something else. `self.config_name`?
I know this will mean much more change than this PR is currently doing.

josephhardinee replied:
I made this change (check notes on most recent commit for more of a sense of scope).

guillaumeeb commented on dask_jobqueue/pbs.py:
```diff
@@ -55,7 +55,9 @@ class PBSCluster(JobQueueCluster):
     cancel_command = 'qdel'
     scheduler_name = 'pbs'
```
You didn't removed scheduler_name here.

josephhardinee replied:
This should be fixed in newest commits on the PR.

guillaumeeb commented:
You're right, but how do we pass this value from PBSCluster to JobQueueCluster constructor? We need to declare the kwarg in JobQueueCluster.__init__() with None as a default too?

mrocklin commented:
Yes, I think that that would work

On Mon, Nov 26, 2018 at 4:40 PM Guillaume Eynard-Bontemps <
notifications@github.com> wrote:

> You're right, but how do we pass this value from PBSCluster to
> JobQueueCluster constructor? We need to declare the kwarg in
> JobQueueCluster.*init*() with None as a default too?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask-jobqueue/pull/204#issuecomment-441809331>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASszAg3Fp3hAORmiQs8lgsrdQODNhUIks5uzF_egaJpZM4YzQG5>
> .
>

josephhardinee commented:
I added another commit that I think addresses all of the points. Commit log has messages but essentially 
1. Changed to config name and it replaces self.scheduler name, but is no longer class variable.
2. Added default kwarg to JobQueueCluster constructor with it set to None. 
3. Added tests for each scheduler (Although they feel a little hacky, having to set all of the parameters). 

In particular if you could double check that 
1. The tests meet your standards and
2. The change to config_name handling on L189 of core.py is correct

mrocklin commented:
This looks good to me.  Thank you for the implementation and the tests @josephhardinee . Merging.

The duplication in tests is interesting.  At some point we might consider making a unittest-style `TestCase` that tests a bunch of standard things for each of the Cluster classes.  I'll raise a separate issue as a followup.

mrocklin commented:
Ah, was already raised as https://github.com/dask/dask-jobqueue/issues/8

guillaumeeb commented:
Thanks for that @josephhardinee!
